logic for deprecated images

Gkrumbach07 opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- Closes: #123 -->
closes: https://github.com/opendatahub-io/odh-dashboard/issues/1370

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->
Added a new annotation to workbenches called `opendatahub.io/image-display-name` which stores the display name of the selected image. this gets used when displaying the image on the notebook table if the image itself could not be found. if it is not found it will also include the wording `(deprecated)`.

When you got to edit this workbench with a deprecated image. the image will be _unselected_ and require you to select a new one

<img width="547" alt="image" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/fe28f22b-26e2-421e-a4a2-d4b15f505028">
<img width="421" alt="image" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/a1c394a3-f2de-4bc0-9c2c-ebb1f63aa14c">
<img width="423" alt="image" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/dbe8125d-7a8b-42b2-aad9-9e0bd453ba2d">


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

1. create a custom notebook image
2. create a workbench with that custom image
3. delete the custom image
4. verify that deprecated wording is on the workbench table
5. verify that on edit, the image selection is required
6. remove the annotation (`opendatahub.io/image-display-name`) from the notebook resource, and notice that the image becomes `unknown` on the table row

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->
I added tests to check the use cases of unknown and deprecated. In doing this i felt the need to implement a few helpers for mocks. namely a `opts: RecursivePartial<k8sKindType>` and then lodash _.merge the base mock with the opts. this gives you the option for fine grained control without having to make a large amount of of options to feed in.

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

Gkrumbach07 commented:
### UX context
This bug fixes the edge case where a user has a already created workbench, but the image tag or the entire image is removed from ODH. an example would be that an admin deleted a custom notebook image but a workbench was using it. In this cases

@kywalker-rh @xianli123 See screenshots for how it would look in the row. For edit, the image selector appears as empty and you are required to select a new image.

@kaedward In the issue you signed off on the wording `Image Name (deprecated)`, but just a fyi in case it looks odd to you

DaoDaoNoCode commented on frontend/src/api/k8s/notebooks.ts:
```diff
@@ -94,13 +96,14 @@ const assembleNotebook = (
         'notebooks.opendatahub.io/last-image-selection': imageSelection,
         'notebooks.opendatahub.io/inject-oauth': 'true',
         'opendatahub.io/username': username,
+        'opendatahub.io/image-display-name': displayName,
```
Should this be the image name here? Why it's using the notebook display name?

Gkrumbach07 replied:
yep you are right. pushed a fix

DaoDaoNoCode commented:
I found an issue that's not caused by your PR, but it will affect the result of your PR so you may want to update it as well here.
In `useNotebookImage` hook, it gets the `tagSoftware` [here](https://github.com/opendatahub-io/odh-dashboard/blob/main/frontend/src/pages/projects/screens/detail/notebooks/useNotebookImage.ts#L30), however, this is the wrong function to use. We should use `getImageVersionSoftwareString` instead of `getRelatedVersionDescription`. When there is only one version for an image, these 2 functions get the same result. However, if there are more than one version, the second function (`getRelatedVersionDescription`) could return an `undefined` because it's only used to show the software string in the image selection dropdown.
So it you want to get a correct and exact software string of a specific image version, you might want to replace it with `getImageVersionSoftwareString(imageVersion)`.
FYI @Gkrumbach07 because if you don't update this hook here, you will get `(deprecated)` tag for those multi-version images.

Gkrumbach07 commented:
@DaoDaoNoCode
> because if you don't update this hook here, you will get (deprecated) tag for those multi-version images.

I saw this in testing so i used an or operator here to prevent the text from only showing `(deprecated)`. It will fall back to using the display name or in the last case `Unknown`
``` tsx
const imageDisplayName =
    notebookImage?.imageName ||
    obj.notebook.metadata.annotations['opendatahub.io/image-display-name'] ||
    'Unknown';
```

However that doesn't fix the case where deprecated will show even if it is not because tag software is undefined. I updated how tag software is fetched based on your comment and adjusted how deprecated is calculated

andrewballantyne commented:
Needs a rebase

christianvogt commented on frontend/src/pages/projects/screens/detail/notebooks/NotebookTableRow.tsx:
```diff
@@ -37,6 +37,12 @@ const NotebookTableRow: React.FC<NotebookTableRowProps> = ({
 
+  const imageDisplayName =
+    notebookImage?.imageName ||
+    obj.notebook.metadata.annotations['opendatahub.io/image-display-name'] ||
```
Our types must be wrong because `annotations` is not a required field in k8s. It's possible to create a notebook via the cli or the console and not have any annotations.
```suggestion
    obj.notebook.metadata.annotations?.['opendatahub.io/image-display-name'] ||
```

andrewballantyne replied:
If types are wrong, this change needs to come with a corresponding type change too. K8s says no to a lot of things that contextually we may know more about the use-case. Maybe Notebooks always have annotations.

Since this is not something we can guarantee on our side, we probably should favour the state of stronger types as that's a runtime error waiting to happen.

Gkrumbach07 replied:
add the `?`

christianvogt replied:
@Gkrumbach07 
Yes add the `?` but also it comes back to the underlying type: `NotebookKind`. It also needs to be updated. `annotations` and `labels` are optional in k8s resources and should be reflected as such in the type.

Gkrumbach07 replied:
I made those optional too. should be updated

christianvogt commented on frontend/src/api/k8s/notebooks.ts:
```diff
@@ -106,6 +107,9 @@ const assembleNotebook = (
         'notebooks.opendatahub.io/last-image-selection': imageSelection,
         'notebooks.opendatahub.io/inject-oauth': 'true',
         'opendatahub.io/username': username,
+        'opendatahub.io/image-display-name': image.imageStream
+          ? getImageStreamDisplayName(image.imageStream)
+          : '',
```
I don't think you should add this annotation at all if it's going to be blank.

christianvogt commented on frontend/src/__tests__/integration/pages/projects/ProjectDetails.spec.ts:
```diff
@@ -29,3 +29,29 @@ test('Non-empty project', async ({ page }) => {
+  // wait for page to load
+  await page.waitForSelector('text=Test Notebook');
+
+  await page.goto(
+    'http://localhost:6006/iframe.html?args=&id=tests-integration-pages-projects-projectdetails--unknown-image&viewMode=story',
+  );
```
Why call `page.goto` again?
Should be able to remove this.

Gkrumbach07 commented:
### Update
- I have added a new state for `disabled`: This required me add an additional `includeDisabled` option for getting imagestreams because we didnt do that by default.
- on edit of a workbench using a disabled image, I allow the user to make edits without forcing them to select a new image: This was a bit odd to do, but i just drilled down an additional prop called `oringalImage` which will track the disabled image option in the select field for the spawner page

These changes came from conversations in an other issue that is _merging_ with this one

context on the merge: https://github.com/opendatahub-io/odh-dashboard/issues/1520#issuecomment-1747665470

xianli123 commented:
@Gkrumbach07 Thanks for invloving me. I left some comments in the issue #1520 , please refer to [my comments](https://github.com/opendatahub-io/odh-dashboard/issues/1520#issuecomment-1751662656).

DaoDaoNoCode commented:
When the notebook image is deleted, should we also disable the toggle button so the users cannot toggle it anymore? It doesn't make sense when we tell the user they need to select another image to start the notebook while we still allow them to start it and get `Error: ImagePullBackOff` later.

<img width="1222" alt="Screenshot 2023-11-02 at 10 43 03 AM" src="https://github.com/opendatahub-io/odh-dashboard/assets/37624318/0d62b17c-588f-41cc-9394-b29f522baeea">

DaoDaoNoCode commented on frontend/src/pages/projects/screens/detail/notebooks/useNotebookImageData.ts:
```diff
@@ -1,52 +1,94 @@
-      return [null, true];
+      return [
+        {
+          loaded,
```
This value should be removed? Looks like the type check is not catching this 🤔

DaoDaoNoCode commented on frontend/src/pages/projects/screens/detail/notebooks/useNotebookImageData.ts:
```diff
@@ -1,52 +1,94 @@
+      {
+        imageStream,
+        imageVersion,
+        loaded,
```
Same here, remove it, I think you are using `loaded` outside of the `data` object.

DaoDaoNoCode reviewed (commented):
Overall it's working as expected. Just left some small comments as well as the question above. Otherwise it looks good!

kywalker-rh commented:
> When the notebook image is deleted, should we also disable the toggle button so the users cannot toggle it anymore? It doesn't make sense when we tell the user they need to select another image to start the notebook while we still allow them to start it and get `Error: ImagePullBackOff` later.
> 
> <img alt="Screenshot 2023-11-02 at 10 43 03 AM" width="1222" src="https://user-images.githubusercontent.com/37624318/280027021-0d62b17c-588f-41cc-9394-b29f522baeea.png">

I think that's a great idea... good catch @DaoDaoNoCode !

DaoDaoNoCode reviewed (approved):
/lgtm

Gkrumbach07 commented:
@christianvogt this PR is ready for your final review

christianvogt commented on frontend/src/pages/projects/screens/spawner/useImageStreams.ts:
```diff
@@ -4,14 +4,15 @@ import { getNotebookImageStreams } from '~/api';
   React.useEffect(() => {
     if (namespace) {
-      getNotebookImageStreams(namespace)
+      getNotebookImageStreams(namespace, includeDisabled)
```
If an error occurs when `includeDisabled` changes and we have previously loaded data, `setImageStreams` will remain the old value as the `catch` clause does not reset the data. Also I don't think we should be setting `loaded` to `true` if we have an error. The typical pattern here sets `loaded` to `true` when data is available. Loaded state should be unaffected by error state.

Gkrumbach07 replied:
I can however set the state back to an empty array to stop previous loaded data from lingering on error

Gkrumbach07 replied:
actually i am just going to switch this over to useFetchState if i can so it uses that common logic

christianvogt commented on frontend/src/pages/projects/screens/detail/notebooks/useNotebookImageData.ts:
```diff
@@ -1,52 +1,92 @@
-    if (!loaded) {
-      return [null, false];
+    if (loadError) {
+      return [null, true, loadError];
```
I don't think loaded should be set to `true` here because this pattern would suggests that data is available.
```suggestion
      return [null, false, loadError];
```

Gkrumbach07 replied:
See comment above

christianvogt commented on frontend/src/pages/projects/screens/detail/notebooks/useNotebookImageData.ts:
```diff
@@ -1,52 +1,92 @@
-      return [null, false];
+    if (!notebook || !loaded) {
+      return [null, false, undefined];
     }
 
-    if (!loaded) {
-      return [null, false];
+    if (loadError) {
+      return [null, true, loadError];
```
I think you should handle error state before loaded state or combine all together if you allow error state to occur when loaded is false.

Gkrumbach07 replied:
now handling like this
```tsx
    if (!loaded || !notebook) {
      return [null, false, loadError];
    }
```

christianvogt commented on frontend/src/pages/projects/screens/detail/notebooks/useNotebookImage.ts:
```diff
@@ -2,35 +2,52 @@ import { NotebookKind } from '~/k8sTypes';
+  | [notebookImage: NotebookImage, loaded: true, loadError: undefined] => {
+  const [data, loaded, loadError] = useNotebookImageData(notebook);
+
+  if (!notebook || !loaded) {
+    return [null, false, undefined];
+  }
+  if (loadError) {
+    return [null, true, loadError];
```
Similar comment elsewhere:
```suggestion
    return [null, false, loadError];
```

christianvogt replied:
Should handle error state before loaded state. But likely can combine all together if you allow error state to occur when loaded is `false`.

christianvogt commented on frontend/src/pages/projects/screens/detail/notebooks/useNotebookImage.ts:
```diff
@@ -2,35 +2,52 @@ import { NotebookKind } from '~/k8sTypes';
-): [notebookImage: NotebookImage | null, loaded: boolean] => {
-  const [imageData, loaded] = useNotebookImageData(notebook);
+):
+  | [notebookImage: null, loaded: false, loadError: undefined]
+  | [notebookImage: null, loaded: true, loadError: Error]
+  | [notebookImage: NotebookImage, loaded: true, loadError: undefined] => {
```
The typical pattern suggests that `loaded=true` implies data is available, but error state is not related to loaded state. In more generic cases, error can be present while data is available if an error occurred on an update.
```suggestion
  | [notebookImage: null, loaded: false, loadError: Error | undefined]
  | [notebookImage: NotebookImage, loaded: true, loadError: undefined] => {
```

christianvogt commented on frontend/src/pages/projects/screens/detail/notebooks/types.ts:
```diff
@@ -0,0 +1,42 @@
+};
+
+export type NotebookImageData =
+  | [data: null, loaded: false, loadError: undefined]
+  | [data: null, loaded: true, loadError: Error]
```
I see that this traces down to `useImageSteams` which sets `loaded` to `true` along with an error. So the logic around this tyoe would work but I believe the pattern is to typically return an error with any value of `loaded`. `useFetchState` can return a subsequent error on refresh while `loaded` was previously set to true.
```suggestion
  | [data: null, loaded: false, loadError: Error | undefined]
```

Gkrumbach07 commented:
@christianvogt i switched the logic around to use useFetchState and have loaded = true to mean there is always some data defined

christianvogt commented:
latest changes to use `useFetchState` look good.
/lgtm
/approve
